File: microService/get_all_teams/get_all_teams.py
import requests
from typing import List


def get_teams(
        state_key: str,
        api_key: str,
        max_page: int = 20,
        url: str = "https://www.thebluealliance.com/api/v3/teams/{}"
) -> List[str]:
    """
    Get all teams in a state

    :param state_key: A team's state to filter based on. You can also pass in `*` to just return everything.
    :param api_key: API Key to use to query the base URL
    :param max_page: The maximum page to stop at.
    :param url: The base URL to query at.
    :return: "List of team numbers"
    """
    team_keys = []
    for page in range(0, max_page):
        response = requests.get(url.format(page), headers={"X-TBA-Auth-Key": api_key})
        if response.status_code == 200:
            teams = response.json()
            for team in teams:
                if state_key != "*":
                    if team.get("state_prov") == state_key:
                        team_keys.append(team.get("team_number"))
                else:
                    team_keys.append(team.get("team_number"))
        else:
            pass
    return team_keys


def get_team_competing(
        event_key: str,
        api_key: str,
        max_page: int = 20,
        url: str = "https://www.thebluealliance.com/api/v3/event/{}/teams"
) -> List[str]:
    """
    Get all teams in a state

    :param event_key: A team's state to filter based on. You can also pass in `*` to just return everything.
    :param api_key: API Key to use to query the base URL
    :param max_page: The maximum page to stop at.
    :param url: The base URL to query at.
    :return: "List of team numbers"
    """
    response = requests.get(url.format(event_key), headers={"X-TBA-Auth-Key": api_key})
    teams = []
    if response.status_code == 200:
        team = response.json()
        for team in team:
            if team.get("team_number"):
                teams.append(str(team.get("team_number")))
        return ", ".join(teams)
    else:
        return []


"""
Example Usage
x = get_teams("State", "API_KEY")
print(x)
"""

File: microService/get_all_teams/get_all_teams.rs
use reqwest::header::HeaderMap;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fmt;
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize)]
struct Team {
    team_number: String,
    state_prov: Option<String>,
}

fn get_teams(
    state_key: &str,
    api_key: &str,
    max_page: i32,
    url: &str,
) -> Result<Vec<String>, Box<dyn Error>> {
    let client = Client::new();
    let mut headers = HeaderMap::new();
    headers.insert(
        "X-TBA-Auth-Key",
        api_key.parse().expect("Failed to parse api key header value"),
    );
    let mut team_keys = Vec::new();
    for page in 0..max_page {
        let page_url = format!("{}{}", url, page);
        let response = client.get(&page_url).headers(headers.clone()).send()?;
        if response.status().is_success() {
            let teams: Vec<Team> = response.json()?;
            for team in teams {
                if state_key != "*" {
                    if let Some(state_prov) = team.state_prov {
                        if state_prov == state_key {
                            team_keys.push(team.team_number);
                        }
                    }
                } else {
                    team_keys.push(team.team_number);
                }
            }
        }
    }
    Ok(team_keys)
}

fn get_team_competing(
    event_key: &str,
    api_key: &str,
    max_page: i32,
    url: &str,
) -> Result<String, Box<dyn Error>> {
    let client = Client::new();
    let mut headers = HeaderMap::new();
    headers.insert(
        "X-TBA-Auth-Key",
        api_key.parse().expect("Failed to parse api key header value"),
    );
    let mut teams = Vec::new();
    let response = client.get(&format!(url, event_key)).headers(headers).send()?;
    if response.status().is_success() {
        let team_list: Vec<HashMap<String, String>> = response.json()?;
        for team in team_list {
            if let Some(team_number) = team.get("team_number") {
                teams.push(team_number.to_string());
            }
        }
        Ok(teams.join(", "))
    } else {
        Ok("".to_string())
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let x = get_teams("State", "API_KEY", 20, "https://www.thebluealliance.com/api/v3/teams/")?;
    println!("{:?}", x);
    Ok(())
}
